File: signet/parse.py
"""Sig-Net Protocol Framework - Packet Parsing.

Receiver-side parsing: CoAP header/option parsing, custom option extraction,
TLV parsing, HMAC verification. Complements send.py.
"""

from __future__ import annotations

import hmac as _hmac
import struct

from .constants import (
    COAP_OPTION_URI_PATH,
    HMAC_SHA256_LENGTH,
    MAX_DMX_SLOTS,
    SENDER_ID_LENGTH,
    SIGNET_OPTION_HMAC,
    SIGNET_OPTION_MFG_CODE,
    SIGNET_OPTION_SECURITY_MODE,
    SIGNET_OPTION_SENDER_ID,
    SIGNET_OPTION_SESSION_ID,
    SIGNET_OPTION_SEQ_NUM,
    TID,
)
from .exceptions import (
    BufferTooSmallError,
    HMACFailedError,
    InvalidOptionError,
    InvalidPacketError,
)
from .security import build_hmac_input, calculate_hmac
from .types import CoAPHeader, SigNetOptions, TLVBlock


class PacketReader:
    """Cursor-based reader over a bytes buffer."""

    def __init__(self, data: bytes) -> None:
        self._data = data
        self._pos = 0

    @property
    def position(self) -> int:
        return self._pos

    @property
    def remaining(self) -> int:
        return len(self._data) - self._pos

    def can_read(self, count: int) -> bool:
        return self._pos + count <= len(self._data)

    def read_byte(self) -> int:
        if not self.can_read(1):
            raise BufferTooSmallError("Cannot read byte: buffer exhausted")
        val = self._data[self._pos]
        self._pos += 1
        return val

    def read_uint16(self) -> int:
        if not self.can_read(2):
            raise BufferTooSmallError("Cannot read uint16: buffer exhausted")
        val = (self._data[self._pos] << 8) | self._data[self._pos + 1]
        self._pos += 2
        return val

    def read_uint32(self) -> int:
        if not self.can_read(4):
            raise BufferTooSmallError("Cannot read uint32: buffer exhausted")
        val = (
            (self._data[self._pos] << 24)
            | (self._data[self._pos + 1] << 16)
            | (self._data[self._pos + 2] << 8)
            | self._data[self._pos + 3]
        )
        self._pos += 4
        return val

    def read_bytes(self, count: int) -> bytes:
        if not self.can_read(count):
            raise BufferTooSmallError(f"Cannot read {count} bytes: buffer exhausted")
        val = self._data[self._pos : self._pos + count]
        self._pos += count
        return val

    def skip(self, count: int) -> None:
        if not self.can_read(count):
            raise BufferTooSmallError(f"Cannot skip {count} bytes: buffer exhausted")
        self._pos += count

    def peek_byte(self) -> int:
        if self._pos >= len(self._data):
            raise BufferTooSmallError("Cannot peek: buffer exhausted")
        return self._data[self._pos]

    @property
    def current_ptr(self) -> bytes:
        return self._data[self._pos :]


def parse_coap_header(reader: PacketReader) -> CoAPHeader:
    """Parse a 4-byte CoAP header."""
    byte0 = reader.read_byte()
    code = reader.read_byte()
    message_id = reader.read_uint16()
    return CoAPHeader(
        version=(byte0 >> 6) & 0x03,
        type_=(byte0 >> 4) & 0x03,
        token_length=byte0 & 0x0F,
        code=code,
        message_id=message_id,
    )


def skip_token(reader: PacketReader, token_length: int) -> None:
    """Skip over CoAP token bytes."""
    if token_length > 0:
        reader.skip(token_length)


def parse_coap_option(reader: PacketReader, prev_option: int) -> tuple[int, bytes]:
    """Parse a single CoAP option.

    Returns (option_number, option_value).
    Raises InvalidPacketError if payload marker (0xFF) is encountered.
    """
    header_byte = reader.peek_byte()
    if header_byte == 0xFF:
        raise InvalidPacketError("Payload marker encountered")

    reader.read_byte()  # consume the byte we peeked

    delta = (header_byte >> 4) & 0x0F
    length = header_byte & 0x0F

    # Extended delta
    if delta == 13:
        delta = 13 + reader.read_byte()
    elif delta == 14:
        delta = 269 + reader.read_uint16()
    elif delta == 15:
        raise InvalidPacketError("Reserved delta value 15")

    # Extended length
    if length == 13:
        length = 13 + reader.read_byte()
    elif length == 14:
        length = 269 + reader.read_uint16()
    elif length == 15:
        raise InvalidPacketError("Reserved length value 15")

    option_num = prev_option + delta
    value = reader.read_bytes(length) if length > 0 else b""

    return option_num, value


def extract_uri_string(reader: PacketReader) -> str:
    """Rebuild URI string from Uri-Path options.

    Returns the full URI (e.g., "/sig-net/v1/level/517").
    Stops when a non-Uri-Path option or payload marker is encountered.
    """
    segments: list[str] = []
    current_option = 0

    while reader.remaining > 0:
        try:
            option_num, value = parse_coap_option(reader, current_option)
        except InvalidPacketError:
            # Hit payload marker
            break

        if option_num == COAP_OPTION_URI_PATH:
            segments.append(value.decode("ascii"))
            current_option = option_num
        elif option_num > COAP_OPTION_URI_PATH:
            # Past Uri-Path options — stop
            break
        else:
            current_option = option_num

    return "/" + "/".join(segments) if segments else "/"


def parse_signet_options(reader: PacketReader) -> SigNetOptions:
    """Parse all 6 SigNet custom options (2076-2236).

    Reader should be positioned after CoAP header + token.
    This parses through all options (including Uri-Path) to find SigNet options.
    """
    options = SigNetOptions()
    current_option = 0
    found = {
        "security_mode": False,
        "sender_id": False,
        "mfg_code": False,
        "session_id": False,
        "seq_num": False,
        "hmac": False,
    }

    while reader.remaining > 0:
        try:
            option_num, value = parse_coap_option(reader, current_option)
        except InvalidPacketError:
            # Payload marker
            break

        if option_num == SIGNET_OPTION_SECURITY_MODE:
            if len(value) != 1:
                raise InvalidOptionError("Security-Mode must be 1 byte")
            options.security_mode = value[0]
            found["security_mode"] = True

        elif option_num == SIGNET_OPTION_SENDER_ID:
            if len(value) != SENDER_ID_LENGTH:
                raise InvalidOptionError(f"Sender-ID must be {SENDER_ID_LENGTH} bytes")
            options = SigNetOptions(
                security_mode=options.security_mode,
                sender_id=value,
                mfg_code=options.mfg_code,
                session_id=options.session_id,
                seq_num=options.seq_num,
                hmac=options.hmac,
            )
            found["sender_id"] = True

        elif option_num == SIGNET_OPTION_MFG_CODE:
            if len(value) != 2:
                raise InvalidOptionError("Mfg-Code must be 2 bytes")
            options.mfg_code = (value[0] << 8) | value[1]
            found["mfg_code"] = True

        elif option_num == SIGNET_OPTION_SESSION_ID:
            if len(value) != 4:
                raise InvalidOptionError("Session-ID must be 4 bytes")
            options.session_id = struct.unpack("!I", value)[0]
            found["session_id"] = True

        elif option_num == SIGNET_OPTION_SEQ_NUM:
            if len(value) != 4:
                raise InvalidOptionError("Seq-Num must be 4 bytes")
            options.seq_num = struct.unpack("!I", value)[0]
            found["seq_num"] = True

        elif option_num == SIGNET_OPTION_HMAC:
            if len(value) != HMAC_SHA256_LENGTH:
                raise InvalidOptionError(f"HMAC must be {HMAC_SHA256_LENGTH} bytes")
            options = SigNetOptions(
                security_mode=options.security_mode,
                sender_id=options.sender_id,
                mfg_code=options.mfg_code,
                session_id=options.session_id,
                seq_num=options.seq_num,
                hmac=value,
            )
            found["hmac"] = True

        current_option = option_num

    if not found["security_mode"]:
        raise InvalidOptionError("Missing Security-Mode option")

    # For normal packets (not beacons), all options required
    if options.security_mode != 0xFF:
        missing = [k for k, v in found.items() if not v]
        if missing:
            raise InvalidOptionError(f"Missing required options: {', '.join(missing)}")

    return options


def parse_tlv_block(reader: PacketReader) -> TLVBlock:
    """Parse a single TLV block from the payload."""
    type_id = reader.read_uint16()
    length = reader.read_uint16()
    value = reader.read_bytes(length) if length > 0 else b""
    return TLVBlock(type_id=type_id, value=value)


def parse_tid_level(tlv: TLVBlock) -> bytes:
    """Extract DMX level data from a TID_LEVEL TLV block."""
    if tlv.type_id != TID.LEVEL:
        raise InvalidPacketError(f"Expected TID_LEVEL (0x{TID.LEVEL:04X}), got 0x{tlv.type_id:04X}")
    if not (1 <= tlv.length <= MAX_DMX_SLOTS):
        raise InvalidPacketError(f"TID_LEVEL length {tlv.length} out of range 1-{MAX_DMX_SLOTS}")
    return tlv.value


def verify_packet_hmac(
    uri_string: str,
    options: SigNetOptions,
    payload: bytes,
    role_key: bytes,
) -> bool:
    """Verify packet HMAC using constant-time comparison.

    Returns True if HMAC matches, raises HMACFailedError otherwise.
    """
    computed = calculate_hmac(uri_string, options, payload, role_key)
    if not _hmac.compare_digest(computed, options.hmac):
        raise HMACFailedError("HMAC verification failed")
    return True

File: src/parse.rs
//==============================================================================
// Sig-Net Protocol Framework - Packet Parsing Functions (Rust)
//==============================================================================
//
// Copyright (c) 2026 Singularity (UK) Ltd.
// License: MIT
//
// Parsing and verification for received Sig-Net packets. CoAP header/option
// parsing, custom option extraction, URI rebuild, TLV parsing, HMAC
// verification. Complements the send module.
//==============================================================================

use crate::constants::*;
use crate::crypto;
use crate::security;
use crate::types::*;

/// PacketReader - helper for reading from a const buffer.
pub struct PacketReader<'a> {
    buffer: &'a [u8],
    position: usize,
}

impl<'a> PacketReader<'a> {
    pub fn new(buffer: &'a [u8]) -> Self {
        PacketReader {
            buffer,
            position: 0,
        }
    }

    pub fn get_position(&self) -> usize {
        self.position
    }

    pub fn get_remaining(&self) -> usize {
        self.buffer.len() - self.position
    }

    pub fn can_read(&self, bytes: usize) -> bool {
        (self.position + bytes) <= self.buffer.len()
    }

    pub fn read_byte(&mut self) -> Option<u8> {
        if !self.can_read(1) {
            return None;
        }
        let value = self.buffer[self.position];
        self.position += 1;
        Some(value)
    }

    /// Read u16 in network byte order (big-endian).
    pub fn read_u16(&mut self) -> Option<u16> {
        if !self.can_read(2) {
            return None;
        }
        let value =
            ((self.buffer[self.position] as u16) << 8) | (self.buffer[self.position + 1] as u16);
        self.position += 2;
        Some(value)
    }

    /// Read u32 in network byte order (big-endian).
    pub fn read_u32(&mut self) -> Option<u32> {
        if !self.can_read(4) {
            return None;
        }
        let value = ((self.buffer[self.position] as u32) << 24)
            | ((self.buffer[self.position + 1] as u32) << 16)
            | ((self.buffer[self.position + 2] as u32) << 8)
            | (self.buffer[self.position + 3] as u32);
        self.position += 4;
        Some(value)
    }

    pub fn read_bytes(&mut self, count: usize) -> Option<&'a [u8]> {
        if !self.can_read(count) {
            return None;
        }
        let slice = &self.buffer[self.position..self.position + count];
        self.position += count;
        Some(slice)
    }

    pub fn skip(&mut self, count: usize) -> bool {
        if !self.can_read(count) {
            return false;
        }
        self.position += count;
        true
    }

    pub fn peek_byte(&self) -> Option<u8> {
        if self.position >= self.buffer.len() {
            return None;
        }
        Some(self.buffer[self.position])
    }

    pub fn get_current_slice(&self) -> &'a [u8] {
        &self.buffer[self.position..]
    }
}

/// Parse CoAP header from the reader.
pub fn parse_coap_header(reader: &mut PacketReader) -> Result<CoAPHeader, i32> {
    let byte0 = reader.read_byte().ok_or(SIGNET_ERROR_BUFFER_TOO_SMALL)?;
    let byte1 = reader.read_byte().ok_or(SIGNET_ERROR_BUFFER_TOO_SMALL)?;
    let message_id = reader.read_u16().ok_or(SIGNET_ERROR_BUFFER_TOO_SMALL)?;

    Ok(CoAPHeader {
        version_type_tkl: byte0,
        code: byte1,
        message_id,
    })
}

/// Skip CoAP token bytes.
pub fn skip_token(reader: &mut PacketReader, token_length: u8) -> i32 {
    if token_length == 0 {
        return SIGNET_SUCCESS;
    }
    if !reader.skip(token_length as usize) {
        return SIGNET_ERROR_BUFFER_TOO_SMALL;
    }
    SIGNET_SUCCESS
}

/// Parsed CoAP option result.
pub struct ParsedOption<'a> {
    pub option_num: u16,
    pub option_value: &'a [u8],
}

/// Parse a single CoAP option using delta encoding.
pub fn parse_coap_option<'a>(
    reader: &mut PacketReader<'a>,
    prev_option: u16,
) -> Result<ParsedOption<'a>, i32> {
    // Peek at next byte to check for payload marker
    let header_byte = reader.peek_byte().ok_or(SIGNET_ERROR_BUFFER_TOO_SMALL)?;

    if header_byte == 0xFF {
        return Err(SIGNET_ERROR_INVALID_PACKET);
    }

    // Read option header
    let header_byte = reader.read_byte().ok_or(SIGNET_ERROR_BUFFER_TOO_SMALL)?;

    let mut delta = ((header_byte >> 4) & 0x0F) as u16;
    let mut length = (header_byte & 0x0F) as u16;

    // Handle extended delta
    match delta {
        13 => {
            let ext = reader.read_byte().ok_or(SIGNET_ERROR_BUFFER_TOO_SMALL)?;
            delta = 13 + ext as u16;
        }
        14 => {
            let ext = reader.read_u16().ok_or(SIGNET_ERROR_BUFFER_TOO_SMALL)?;
            delta = 269 + ext;
        }
        15 => return Err(SIGNET_ERROR_INVALID_PACKET),
        _ => {}
    }

    // Handle extended length
    match length {
        13 => {
            let ext = reader.read_byte().ok_or(SIGNET_ERROR_BUFFER_TOO_SMALL)?;
            length = 13 + ext as u16;
        }
        14 => {
            let ext = reader.read_u16().ok_or(SIGNET_ERROR_BUFFER_TOO_SMALL)?;
            length = 269 + ext;
        }
        15 => return Err(SIGNET_ERROR_INVALID_PACKET),
        _ => {}
    }

    let option_num = prev_option + delta;

    let option_value = if length > 0 {
        reader
            .read_bytes(length as usize)
            .ok_or(SIGNET_ERROR_BUFFER_TOO_SMALL)?
    } else {
        &[]
    };

    Ok(ParsedOption {
        option_num,
        option_value,
    })
}

/// Extract URI string from CoAP Uri-Path options.
///
/// Rebuilds the full URI string (e.g., "/sig-net/v1/level/517").
pub fn extract_uri_string(reader: &mut PacketReader) -> Result<String, i32> {
    let mut uri = String::from("/");
    let mut current_option: u16 = 0;
    let mut first_segment = true;

    loop {
        match parse_coap_option(reader, current_option) {
            Ok(opt) => {
                if opt.option_num == COAP_OPTION_URI_PATH {
                    if !first_segment {
                        uri.push('/');
                    }
                    first_segment = false;

                    let segment = std::str::from_utf8(opt.option_value)
                        .map_err(|_| SIGNET_ERROR_INVALID_PACKET)?;
                    uri.push_str(segment);
                } else if opt.option_num > COAP_OPTION_URI_PATH {
                    break;
                }
                current_option = opt.option_num;
            }
            Err(SIGNET_ERROR_INVALID_PACKET) => {
                // Hit payload marker or end of options
                break;
            }
            Err(e) => return Err(e),
        }
    }

    Ok(uri)
}

/// Parse SigNet custom options (2076-2236) from a packet.
pub fn parse_signet_options(reader: &mut PacketReader) -> Result<SigNetOptions, i32> {
    let mut options = SigNetOptions::default();
    let mut current_option: u16 = 0;

    let mut found_security_mode = false;
    let mut found_sender_id = false;
    let mut found_mfg_code = false;
    let mut found_session_id = false;
    let mut found_seq_num = false;
    let mut found_hmac = false;

    loop {
        match parse_coap_option(reader, current_option) {
            Ok(opt) => {
                match opt.option_num {
                    SIGNET_OPTION_SECURITY_MODE => {
                        if opt.option_value.len() != 1 {
                            return Err(SIGNET_ERROR_INVALID_OPTION);
                        }
                        options.security_mode = opt.option_value[0];
                        found_security_mode = true;
                    }
                    SIGNET_OPTION_SENDER_ID => {
                        if opt.option_value.len() != SENDER_ID_LENGTH {
                            return Err(SIGNET_ERROR_INVALID_OPTION);
                        }
                        options.sender_id.copy_from_slice(opt.option_value);
                        found_sender_id = true;
                    }
                    SIGNET_OPTION_MFG_CODE => {
                        if opt.option_value.len() != 2 {
                            return Err(SIGNET_ERROR_INVALID_OPTION);
                        }
                        options.mfg_code =
                            ((opt.option_value[0] as u16) << 8) | (opt.option_value[1] as u16);
                        found_mfg_code = true;
                    }
                    SIGNET_OPTION_SESSION_ID => {
                        if opt.option_value.len() != 4 {
                            return Err(SIGNET_ERROR_INVALID_OPTION);
                        }
                        options.session_id = ((opt.option_value[0] as u32) << 24)
                            | ((opt.option_value[1] as u32) << 16)
                            | ((opt.option_value[2] as u32) << 8)
                            | (opt.option_value[3] as u32);
                        found_session_id = true;
                    }
                    SIGNET_OPTION_SEQ_NUM => {
                        if opt.option_value.len() != 4 {
                            return Err(SIGNET_ERROR_INVALID_OPTION);
                        }
                        options.seq_num = ((opt.option_value[0] as u32) << 24)
                            | ((opt.option_value[1] as u32) << 16)
                            | ((opt.option_value[2] as u32) << 8)
                            | (opt.option_value[3] as u32);
                        found_seq_num = true;
                    }
                    SIGNET_OPTION_HMAC => {
                        if opt.option_value.len() != HMAC_SHA256_LENGTH {
                            return Err(SIGNET_ERROR_INVALID_OPTION);
                        }
                        options.hmac.copy_from_slice(opt.option_value);
                        found_hmac = true;
                    }
                    _ => {}
                }
                current_option = opt.option_num;
            }
            Err(SIGNET_ERROR_INVALID_PACKET) => {
                // Hit payload marker - done with options
                break;
            }
            Err(e) => return Err(e),
        }
    }

    if !found_security_mode {
        return Err(SIGNET_ERROR_INVALID_OPTION);
    }

    // For normal packets (not beacons), all options are required
    if options.security_mode != 0xFF
        && (!found_sender_id
            || !found_mfg_code
            || !found_session_id
            || !found_seq_num
            || !found_hmac)
    {
        return Err(SIGNET_ERROR_INVALID_OPTION);
    }

    Ok(options)
}

/// Parsed TLV block (owns the value reference into the original buffer).
pub struct ParsedTLVBlock<'a> {
    pub type_id: u16,
    pub length: u16,
    pub value: &'a [u8],
}

/// Parse the next TLV block from the payload.
pub fn parse_tlv_block<'a>(reader: &mut PacketReader<'a>) -> Result<ParsedTLVBlock<'a>, i32> {
    let type_id = reader.read_u16().ok_or(SIGNET_ERROR_BUFFER_TOO_SMALL)?;
    let length = reader.read_u16().ok_or(SIGNET_ERROR_BUFFER_TOO_SMALL)?;

    let value = if length > 0 {
        reader
            .read_bytes(length as usize)
            .ok_or(SIGNET_ERROR_BUFFER_TOO_SMALL)?
    } else {
        &[]
    };

    Ok(ParsedTLVBlock {
        type_id,
        length,
        value,
    })
}

/// Parse TID_LEVEL payload - extract DMX level data.
pub fn parse_tid_level(tlv: &ParsedTLVBlock) -> Result<Vec<u8>, i32> {
    if tlv.type_id != TID_LEVEL {
        return Err(SIGNET_ERROR_INVALID_ARG);
    }
    if tlv.length < 1 || tlv.length > 512 {
        return Err(SIGNET_ERROR_INVALID_PACKET);
    }

    Ok(tlv.value.to_vec())
}

/// Verify packet HMAC using constant-time comparison to prevent timing attacks.
pub fn verify_packet_hmac(
    uri_string: &str,
    options: &SigNetOptions,
    payload: &[u8],
    role_key: &[u8; DERIVED_KEY_LENGTH],
) -> i32 {
    // Build HMAC input buffer
    let hmac_input = match security::build_hmac_input(uri_string, options, payload) {
        Ok(v) => v,
        Err(e) => return e,
    };

    // Calculate expected HMAC
    let computed_hmac = match crypto::hmac_sha256(role_key, &hmac_input) {
        Ok(h) => h,
        Err(e) => return e,
    };

    // Constant-time comparison
    let mut diff: u8 = 0;
    for i in 0..HMAC_SHA256_LENGTH {
        diff |= computed_hmac[i] ^ options.hmac[i];
    }

    if diff != 0 {
        return SIGNET_ERROR_HMAC_FAILED;
    }

    SIGNET_SUCCESS
}
